apps/lib/chain_pusher/solana.go

```diff
@@ -0,0 +1,411 @@
+			continue
+		}
+
+		if accountInfo == nil || len(accountInfo.Value.Data.GetBinary()) == 0 {
+			sci.logger.Debug().Str("assetId", hex.EncodeToString(encodedAssetId[:])).Msg("No value found")
+			continue
+		}
+
+		// Decode the account data
+		decoder := bin.NewBorshDecoder(accountInfo.Value.Data.GetBinary())
+		account := &contract.TemporalNumericValueFeedAccount{}
+		err = account.UnmarshalWithDecoder(decoder)
+		if err != nil {
+			sci.logger.Error().Err(err).Str("account", feedAccount.String()).Msg("Failed to decode account data")
+			continue
+		}
+
+		// Convert to internal format
+		polledVals[encodedAssetId] = InternalStorkStructsTemporalNumericValue{
+			QuantizedValue: account.LatestValue.QuantizedValue.BigInt(),
+			TimestampNs:    account.LatestValue.TimestampNs,
+		}
+	}
+
+	return polledVals, nil
+}
+
+func (sci *SolanaContractInteracter) BatchPushToContract(priceUpdates map[InternalEncodedAssetId]AggregatedSignedPrice) error {
+	// Create a rate limiter: 40 requests per 10 seconds (4 per second)
+	// with a burst of 10 to stay well within the limits
+	limiter := rate.NewLimiter(rate.Every(250*time.Millisecond), 10)
+
+	var wg sync.WaitGroup
+	errChan := make(chan error, len(priceUpdates))
+
+	priceUpdatesBatch := make(map[InternalEncodedAssetId]AggregatedSignedPrice)
+	i := 0
+	for encodedAssetId, priceUpdate := range priceUpdates {
+		priceUpdatesBatch[encodedAssetId] = priceUpdate
+		i++
+		if len(priceUpdatesBatch) == MAX_BATCH_SIZE || i == len(priceUpdates) {
+			batchCopy := make(map[InternalEncodedAssetId]AggregatedSignedPrice, len(priceUpdatesBatch))
+			for k, v := range priceUpdatesBatch {
+				batchCopy[k] = v
+			}
+			wg.Add(1)
+			go func(priceUpdateBatch map[InternalEncodedAssetId]AggregatedSignedPrice) {
+				defer wg.Done()
+				err := limiter.Wait(context.Background())
+				if err != nil {
+					errChan <- fmt.Errorf("rate limiter error: %w", err)
+					return
+				}
+				err = sci.pushLimitedBatchUpdateToContract(priceUpdateBatch)
+				if err != nil {
+					errChan <- fmt.Errorf("failed to push batch: %w", err)
+				}
+			}(batchCopy)
+			priceUpdatesBatch = make(map[InternalEncodedAssetId]AggregatedSignedPrice)
+		}
+	}
+
+	wg.Wait()
+	close(errChan)
+
+	// Collect any errors
+	var errs []error
+	for err := range errChan {
+		errs = append(errs, err)
+	}
+
+	// If there were any errors, return them combined
+	if len(errs) > 0 {
+		return fmt.Errorf("batch push encountered %d errors: %v", len(errs), errs)
+	}
+
+	sci.logger.Info().
+		Int("numUpdates", len(priceUpdates)).
+		Msg("Successfully pushed batch updates to contract")
+
+	return nil
+}
+
+func (sci *SolanaContractInteracter) pushLimitedBatchUpdateToContract(priceUpdates map[InternalEncodedAssetId]AggregatedSignedPrice) error {
+
+	if len(priceUpdates) > MAX_BATCH_SIZE {
+		sci.logger.Error().Msg("Batch size exceeds limit, skipping update")
+		return nil
+	}
+
+	configAccount, _, err := solana.FindProgramAddress(
```

Comment:
I like that the feed address is defined in the init call. Thoughts on doing the same for config and treasuries to save time?